File: src/main/java/com/indico/query/ListWorkflows.java
package com.indico.query;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.api.Response;
import com.indico.Async;
import com.indico.IndicoClient;
import com.indico.ListWorkflowsGraphQLQuery;
import com.indico.Query;
import com.indico.entity.Workflow;

import java.util.ArrayList;
import java.util.List;

public class ListWorkflows implements Query<List<Workflow>> {

    private List<Integer> datasetIds;
    private List<Integer> workflowIds;
    private final IndicoClient client;

    public ListWorkflows(IndicoClient client) {
        this.client = client;
    }

    /**
     * List of dataset ids to filter by
     * @param ids Dataset Ids
     * @return ListWorkflows
     */
    public ListWorkflows datasetIds(List<Integer> ids) {
        this.datasetIds = ids;
        return this;
    }

    /**
     * List of workflow ids to filter by
     * @param ids
     * @return ListWorkflows
     */
    public ListWorkflows workflowIds(List<Integer> ids) {
        this.workflowIds = ids;
        return this;
    }

    /**
     * Queries the server and returns Workflow List
     * @return Workflow List
     */
    @Override
    public List<Workflow> query() {
        ApolloCall<ListWorkflowsGraphQLQuery.Data> apolloCall = this.client.apolloClient.query(ListWorkflowsGraphQLQuery.builder()
                .datasetIds(this.datasetIds)
                .workflowIds(this.workflowIds)
                .build());
        Response<ListWorkflowsGraphQLQuery.Data> response = (Response<ListWorkflowsGraphQLQuery.Data>) Async.executeSync(apolloCall, this.client.config).join();

        List<ListWorkflowsGraphQLQuery.Workflow> wf = response.data().workflows().workflows();
        List<Workflow> workflows = new ArrayList<>();
        wf.forEach(workflow -> workflows.add(new Workflow.Builder()
                .id(workflow.id())
                .name(workflow.name())
                .reviewEnabled(workflow.reviewEnabled())
                .build()));
        return workflows;
    }

    @Override
    public List<Workflow> refresh(List<Workflow> obj) {
        return obj;
    }
}

File: src/main/java/com/indico/query/ListWorkflows.kt
package com.indico.query

import com.expediagroup.graphql.client.jackson.types.OptionalInput
import com.indico.IndicoClient
import com.indico.entity.Workflow
import com.indico.graphql.ListWorkflowsGraphQL
import java.util.ArrayList

class ListWorkflows(private val client: IndicoClient) : Query<List<Workflow>?, ListWorkflowsGraphQL.Result>() {
    private var datasetIds: List<Int>? = null
    private var workflowIds: List<Int>? = null

    /**
     * List of dataset ids to filter by
     * @param ids Dataset Ids
     * @return ListWorkflows
     */
    fun datasetIds(ids: List<Int>?): ListWorkflows {
        datasetIds = ids
        return this
    }

    /**
     * List of workflow ids to filter by
     * @param ids
     * @return ListWorkflows
     */
    fun workflowIds(ids: List<Int>?): ListWorkflows {
        workflowIds = ids
        return this
    }

    /**
     * Queries the server and returns Workflow List
     * @return Workflow List
     */
    override fun query(): List<Workflow> {
        return try {
            val call = ListWorkflowsGraphQL(ListWorkflowsGraphQL.Variables(
                datasetIds = if (datasetIds != null) OptionalInput.Defined(datasetIds) else OptionalInput.Undefined,
                workflowIds = if (workflowIds !=null) OptionalInput.Defined(workflowIds) else OptionalInput.Undefined

            ))
            val response = client.execute(call)
            handleErrors(response)
            val wf = response.data?.workflows?.workflows?: ArrayList()
            val workflows: MutableList<Workflow> = ArrayList()
            wf.forEach { workflow ->
                workflows.add(
                    Workflow.Builder()
                        .id(workflow!!.id!!)
                        .name(workflow.name!!)
                        .reviewEnabled(workflow.reviewEnabled!!)
                        .build()
                )
            }
            workflows
        } catch (ex: InterruptedException) {
            throw RuntimeException("Call to list workflows failed", ex)
        }
    }

    override fun refresh(obj: List<Workflow>?): List<Workflow>? {
        return obj
    }
}